Add support for reprompts if user input does not match template schema

shreyas-goenka opened the pull request:
## Changes
This PR adds retry logic to user input prompts, prompting users again if the value does not match the requirements specified in the bundle template schema.

## Tests
Manually. Here's an example UX. The first prompt expects an integer and the second one a string made only from the letters "defg"

```
shreyas.goenka@THW32HFW6T cli % cli bundle init ~/mlops-stack

Please enter an integer [123]: abc
Validation failed: "abc" is not a integer

Please enter an integer [123]: 123

Please enter a string [dddd]: apple
Validation failed: invalid value for input_root_dir: "apple". Only characters the 'd', 'e', 'f', 'g' are allowed
```

pietern commented on libs/template/config.go:
```diff
@@ -92,63 +92,102 @@ func (c *config) assignDefaultValues(r *renderer) error {
+	}
 
-		description, err := r.executeTemplate(property.Description)
+	for i := 0; i < numRetries; i++ {
```
Why limit retries? User can always Ctrl-C out of it if they're fed up.

shreyas-goenka replied:
Done. We shall bring the power of infinite retries to our customers.

pietern commented on libs/template/config.go:
```diff
@@ -92,63 +92,102 @@ func (c *config) assignDefaultValues(r *renderer) error {
-		// Validate the partial config based on this update
-		if err := c.schema.ValidateInstance(c.values); err != nil {
+		// Validate the partial config based on this new value by the user
+		err = c.schema.ValidateInstance(c.values)
```
This is duplicated in the other prompt function.

shreyas-goenka replied:
Not really duplicated right? For `promptText` this is wrapped in a retry loop.

pietern replied:
So for `promptSelect` we never retry if it fails?

shreyas-goenka replied:
I believe there's no reason to retry `promptSelect` atm. The reason we would like to retry `promptText` is because it's likely the user entered value is of the wrong type or does not match the regex pattern required.

In case of `promptSelect`, An error likely implies something went wrong with the prompting itself, rather than having to do something with the value a user selected.

shreyas-goenka replied:
Done. Moved retry loop to the `promptForValues` function. It's probably good enough to retry on all validation errors. WDYT?

pietern reviewed (commented):
The retry loop should be done in the `promptForValues` function to reduce duplication. The two different prompt functions can return a proper error or a retryable error and depending on the type can the loop decide to retry or not.

pietern commented:
W.r.t. the error message:
> Validation failed: could not parse "abc" as a integer: strconv.ParseInt: parsing "abc": invalid syntax

This exposes internals. A more friendly error message would be:
> Validation failed: "abc" is not an integer.

shreyas-goenka commented:
Output logs after addressing comments:
```
shreyas.goenka@THW32HFW6T playground % cli bundle init ~/mlops-stack

Please enter an integer [123]: abc
Validation failed: "abc" is not an integer

Please enter an integer [123]: 234

Please enter a string [dddd]: lope
Validation failed: "lope" is expected to match regex pattern ^[d-g]+$. Only characters the 'd', 'e', 'f', 'g' are allowed

Please enter a string [dddd]: defg

Select cloud: 
  ▸ azure
    aws
```

Template schema for these properties:
```
  "properties": {
    "input_project_name": {
      "order": 1,
      "type": "integer",
      "default": 123,
      "description": "\nPlease enter an integer"
    },
    "input_root_dir": {
      "order": 2,
      "type": "string",
      "default": "dddd",
      "description": "\nPlease enter a string",
      "pattern": "^[d-g]+$",
      "pattern_match_failure_message": "Only characters the 'd', 'e', 'f', 'g' are allowed"
    },
```

pietern commented on libs/template/config.go:
```diff
@@ -143,11 +144,66 @@ func (c *config) skipPrompt(p jsonschema.Property, r *renderer) (bool, error) {
+
+func (c *config) promptText(name, description, defaultVal string, propertySchema *jsonschema.Schema) error {
+	for {
+		// Get user input. Parse the string back to a value
```
Comment seems wrong? Conversion back to a value happens below.

pietern commented on libs/template/config.go:
```diff
@@ -143,11 +144,66 @@ func (c *config) skipPrompt(p jsonschema.Property, r *renderer) (bool, error) {
+		if err != nil {
+			target := &jsonschema.ParseStringError{}
+			if errors.As(err, target) {
+				cmdio.LogString(c.ctx, fmt.Sprintf("Validation failed: %q is not a %s", target.Value, target.ExpectedType))
```
If we reformat the error here, why do we have the formatting in its `Error()` function?

As this `err != nil` block retries regardless of the type, it can defer to the underlying error for formatting.

pietern commented:
The PR description is out of date from the latest behavior.

andrewnester commented on libs/template/config.go:
```diff
@@ -171,39 +171,50 @@ func (c *config) promptForValues(r *renderer) error {
-				return err
+		// We wrap this function in a retry loop to allow retries when the user
+		// entered value is invalid.
+		for {
```
can you extract this code into separate method for better readability as `promptForValues` became a bit too big to follow

shreyas-goenka replied:
Done :)

andrewnester commented on libs/template/config.go:
```diff
@@ -171,38 +218,22 @@ func (c *config) promptForValues(r *renderer) error {
-			userInput, err = cmdio.Ask(c.ctx, description, defaultVal)
-			if err != nil {
-				return err
+			if errors.As(err, &retriableError{}) {
```
```suggestion
			if !errors.As(err, &retriableError{}) {
				return err
			}
```

shreyas-goenka replied:
Done

andrewnester commented on libs/template/config.go:
```diff
@@ -143,6 +152,44 @@ func (c *config) skipPrompt(p jsonschema.Property, r *renderer) (bool, error) {
 
+func (c *config) promptOnce(property *jsonschema.Schema, name, defaultVal, description string) error {
+	var userInput string
+	var err error
```
Why is it defined at top level?

shreyas-goenka replied:
Needed because we use `userInput, err = ...` and `c.values[name], err = ...`. `:=` is not an option because the first option exists beforehand in both cases. Moved the declaration closer to where it's used to make this clearer.
